src/robot/running/keywordrunner.py

```diff
@@ -179,13 +188,54 @@ def _get_values_for_one_round(self, data):
 
     def _replace_variables(self, data):
         values = self._context.variables.replace_list(data.values)
-        if data.range:
-            values = self._get_range_items(values)
-        if len(values) % len(data.variables) == 0:
+        values = self._transform_items(values)
+        values_per_iteration = self._original_values_per_iteration(data.variables)
+        if len(values) % values_per_iteration == 0:
             return values
+        error_addendum = ""
+        if len(data.variables) != values_per_iteration:
+            # This is a bit awkward, but I'm not sure how else to phrase
+            # the impact of a non-value variable like from ForInEnumerate.
+            error_addendum = " (%d of which matter for this error)" % values_per_iteration
```

Comment:
I think it would be better to have a separate method that raises the exception (or returns the error message) that ForInEnumerate can override.